File: src/otto_trans/core/cache.py
import atexit
import base64
import sqlite3
from pathlib import Path


class Cache:
    _instance = None
    _initialized = False
    db_path = Path.home() / ".cache" / "otto-trans" / "cache.db"

    def __new__(cls, *args, **kwargs):
        if not cls._instance:
            cls._instance = super().__new__(cls, *args, **kwargs)
        return cls._instance

    def __init__(self):
        if self._initialized:
            return
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(str(self.db_path))
        self._initialized = True
        atexit.register(self._conn.close)

    def __del__(self):
        self._conn.close()

    def query(self, key: str, engine: str, src_lang: str, tgt_lang: str) -> str | None:
        cursor = self._conn.cursor()
        try:
            cursor.execute(
                f"SELECT target FROM [{self._b64encode(engine)}] WHERE source = ? AND src_lang = ? AND tgt_lang = ?",
                (
                    self._b64encode(key),
                    self._b64encode(src_lang),
                    self._b64encode(tgt_lang),
                ),
            )
        except sqlite3.OperationalError:
            return None  # 表不存在 = 无缓存
        result = cursor.fetchone()
        return self._b64decode(result[0]) if result else None

    def insert(self, key: str, value: str, engine: str, src_lang: str, tgt_lang: str):
        cursor = self._conn.cursor()
        self._create_table(self._b64encode(engine))
        cursor.execute(
            f"INSERT OR REPLACE INTO [{self._b64encode(engine)}] (source, target, src_lang, tgt_lang) VALUES (?, ?, ?, ?)",
            (
                self._b64encode(key),
                self._b64encode(value),
                self._b64encode(src_lang),
                self._b64encode(tgt_lang),
            ),
        )
        self._conn.commit()

    def _create_table(self, name: str):
        cursor = self._conn.cursor()
        cursor.execute(f"""
            CREATE TABLE IF NOT EXISTS [{name}] (
                source TEXT NOT NULL,
                target TEXT NOT NULL,
                src_lang TEXT NOT NULL,
                tgt_lang TEXT NOT NULL,
                PRIMARY KEY (source, src_lang, tgt_lang)
            )
        """)

    @staticmethod
    def _b64encode(s: str) -> str:
        return base64.b64encode(s.encode(encoding="utf-8")).decode(encoding="utf-8")

    @staticmethod
    def _b64decode(s: str) -> str:
        return base64.b64decode(s.encode(encoding="utf-8")).decode(encoding="utf-8")

    @classmethod
    def reset(cls):
        if cls._instance is not None:
            cls._instance._conn.close()
        cls.db_path.unlink(missing_ok=True)
        if cls._instance is not None:
            cls._instance._conn = sqlite3.connect(str(cls.db_path))

File: otto-trans-java/src/main/java/kg/us/xuan/otto_trans/core/Cache.java
package kg.us.xuan.otto_trans.core;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Base64;

public class Cache {

    public static final Path dbPath = Path.of(System.getProperty("user.home"), ".cache", "otto-trans", "cache.db");

    private static final Cache INSTANCE = new Cache();

    private Connection conn = null;

    private Cache() {
        try {
            Files.createDirectories(dbPath.getParent());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        try {
            Class.forName("org.sqlite.JDBC");
            conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + dbPath);
            conn.setAutoCommit(false);
        } catch (ClassNotFoundException | SQLException e) {
            throw new RuntimeException(e);
        }
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try {
                if (conn != null && !conn.isClosed()) {
                    conn.close();
                }
            } catch (SQLException e) {
                // 无论如何都不抛出异常了
            }
        }));
    }

    public static Cache getInstance() {
        return INSTANCE;
    }

    public String query(String key, String engine, String srcLang, String tgtLang) throws SQLException {
        String result = null;
        String sql = String.format("SELECT target FROM [%s] WHERE source = ? AND src_lang = ? AND tgt_lang = ?",
                b64Encode(engine));
        try (PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setString(1, b64Encode(key));
            stmt.setString(2, b64Encode(srcLang));
            stmt.setString(3, b64Encode(tgtLang));
            try {
                ResultSet rs = stmt.executeQuery();
                if (rs.next()) {
                    result = b64Decode(rs.getString(1));
                }
            } catch (SQLException e) {
                // 表不存在，啥都不干
            }
        }
        return result;
    }

    public void insert(String key, String value, String engine, String srcLang, String tgtLang) throws SQLException {
        createTable(b64Encode(engine));
        String sql = String.format("INSERT OR REPLACE INTO [%s] (source, target, src_lang, tgt_lang) VALUES (?, ?, ?, ?)",
                b64Encode(engine));
        try (PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setString(1, b64Encode(key));
            stmt.setString(2, b64Encode(value));
            stmt.setString(3, b64Encode(srcLang));
            stmt.setString(4, b64Encode(tgtLang));
            stmt.executeUpdate();
        }
        conn.commit();
    }

    private void createTable(String name) throws SQLException {
        String sql = String.format(
                "CREATE TABLE IF NOT EXISTS [%s] (source TEXT NOT NULL, target TEXT NOT NULL, src_lang TEXT NOT NULL, tgt_lang TEXT NOT NULL, PRIMARY KEY (source, src_lang, tgt_lang))",
                name);
        try (PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.execute();
        }
    }

    private String b64Encode(String s) {
        return Base64.getEncoder().encodeToString(s.getBytes(StandardCharsets.UTF_8));
    }

    private String b64Decode(String s) {
        return new String(Base64.getDecoder().decode(s), StandardCharsets.UTF_8);
    }
}
